File: src/Urql.Core/Intermediate/Compiler.cs
using Urql.Core.Diagnostics;
using Urql.Core.Syntax;
using Urql.Core.Syntax.Ast;

namespace Urql.Core.Intermediate;

public static class Compiler
{
    private sealed record PendingPatch(int InstructionIndex, string LabelName, bool IsCall, SourceSpan Span);

    public static IrProgram Compile(ProgramSyntax program, IReadOnlyList<Diagnostic>? parseDiagnostics = null)
    {
        var diagnostics = new DiagnosticBag();
        if (parseDiagnostics is not null)
        {
            diagnostics.AddRange(parseDiagnostics);
        }

        var instructions = new List<IrInstruction>();
        var labels = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
        var pending = new List<PendingPatch>();

        foreach (var line in program.Lines)
        {
            if (line.Label is not null)
            {
                if (!labels.TryAdd(line.Label.Name, instructions.Count))
                {
                    diagnostics.ReportError(
                        CompileDiagnosticCode.DuplicateLabel,
                        $"Duplicate label '{line.Label.Name}'.",
                        line.Label.Span);
                }
            }

            foreach (var statement in line.Statements)
            {
                EmitStatement(statement, instructions, pending);
            }
        }

        for (var i = 0; i < pending.Count; i++)
        {
            var patch = pending[i];
            if (!labels.TryGetValue(patch.LabelName, out var target))
            {
                diagnostics.Report(
                    CompileDiagnosticCode.UnknownLabel,
                    DiagnosticSeverity.Warning,
                    $"Unknown label '{patch.LabelName}'.",
                    patch.Span);
                continue;
            }

            if (patch.IsCall)
            {
                instructions[patch.InstructionIndex] = new CallInstruction(target, patch.Span);
            }
            else
            {
                instructions[patch.InstructionIndex] = new JumpInstruction(target, patch.Span);
            }
        }

        return new IrProgram(instructions, labels, diagnostics.Items);
    }

    private static void EmitStatement(
        StatementSyntax statement,
        List<IrInstruction> instructions,
        List<PendingPatch> pending)
    {
        switch (statement)
        {
            case AssignmentStatementSyntax assign:
                instructions.Add(new AssignInstruction(assign.Name, assign.Expression, false, assign.Span));
                break;

            case InstrStatementSyntax instr:
                instructions.Add(new AssignInstruction(instr.Name, instr.Expression, true, instr.Span));
                break;

            case PrintStatementSyntax print:
                instructions.Add(new PrintInstruction(print.TextExpression, print.AppendNewLine, print.Span));
                break;

            case BtnStatementSyntax btn:
                instructions.Add(new AddButtonInstruction(btn.TargetExpression, btn.CaptionExpression, btn.Span));
                break;

            case PerkillStatementSyntax perk:
                instructions.Add(new PerkillInstruction(perk.Span));
                break;

            case InvkillStatementSyntax invkill:
                instructions.Add(new InvkillInstruction(invkill.ItemExpression, invkill.Span));
                break;

            case InvAddStatementSyntax invAdd:
                instructions.Add(new InvAddInstruction(invAdd.CountExpression, invAdd.ItemExpression, invAdd.Span));
                break;

            case InvRemoveStatementSyntax invRemove:
                instructions.Add(new InvRemoveInstruction(invRemove.CountExpression, invRemove.ItemExpression, invRemove.Span));
                break;

            case EndStatementSyntax end:
                instructions.Add(new ReturnOrHaltInstruction(end.Span));
                break;

            case GotoStatementSyntax g:
            {
                if (TryStaticLabel(g.Target, out var label))
                {
                    var idx = instructions.Count;
                    instructions.Add(new JumpInstruction(-1, g.Span));
                    pending.Add(new PendingPatch(idx, label, false, g.Span));
                }
                else
                {
                    instructions.Add(new JumpDynamicInstruction(g.Target, g.Span));
                }

                break;
            }

            case ProcStatementSyntax p:
            {
                if (TryStaticLabel(p.Target, out var label))
                {
                    var idx = instructions.Count;
                    instructions.Add(new CallInstruction(-1, p.Span));
                    pending.Add(new PendingPatch(idx, label, true, p.Span));
                }
                else
                {
                    instructions.Add(new CallDynamicInstruction(p.Target, p.Span));
                }

                break;
            }

            case IfStatementSyntax ifs:
                EmitIf(ifs, instructions, pending);
                break;

            case UnknownCommandStatementSyntax:
                instructions.Add(new NoOpInstruction(statement.Span));
                break;

            default:
                instructions.Add(new NoOpInstruction(statement.Span));
                break;
        }
    }

    private static void EmitIf(
        IfStatementSyntax ifs,
        List<IrInstruction> instructions,
        List<PendingPatch> pending)
    {
        var jumpFalseIndex = instructions.Count;
        instructions.Add(new JumpIfFalseInstruction(ifs.Condition, -1, ifs.Span));

        foreach (var s in ifs.ThenStatements)
        {
            EmitStatement(s, instructions, pending);
        }

        if (ifs.ElseStatements is { Count: > 0 })
        {
            var jumpEndIndex = instructions.Count;
            instructions.Add(new JumpInstruction(-1, ifs.Span));

            var elseStart = instructions.Count;
            instructions[jumpFalseIndex] = new JumpIfFalseInstruction(ifs.Condition, elseStart, ifs.Span);

            foreach (var s in ifs.ElseStatements)
            {
                EmitStatement(s, instructions, pending);
            }

            instructions[jumpEndIndex] = new JumpInstruction(instructions.Count, ifs.Span);
        }
        else
        {
            instructions[jumpFalseIndex] = new JumpIfFalseInstruction(ifs.Condition, instructions.Count, ifs.Span);
        }
    }

    private static bool TryStaticLabel(ExpressionSyntax expression, out string label)
    {
        switch (expression)
        {
            case IdentifierExpressionSyntax id when !string.IsNullOrWhiteSpace(id.Name):
                label = id.Name;
                return true;

            case StringLiteralExpressionSyntax str when !string.IsNullOrWhiteSpace(str.Value):
                label = str.Value;
                return true;

            default:
                label = string.Empty;
                return false;
        }
    }
}

File: web/packages/core/src/compiler.ts
import type {
  ExpressionSyntax,
  ProgramSyntax,
  StatementSyntax
} from "./ast.js";
import { DiagnosticBag, diagnosticCode, type Diagnostic } from "./diagnostics.js";
import type { SourceSpan } from "./source.js";
import type { IrInstruction, IrProgram } from "./ir.js";

interface PendingPatch {
  instructionIndex: number;
  labelName: string;
  isCall: boolean;
  span: SourceSpan;
}

export class Compiler {
  static compile(program: ProgramSyntax, parseDiagnostics: Diagnostic[] = []): IrProgram {
    const diagnostics = new DiagnosticBag();
    diagnostics.items.push(...parseDiagnostics);

    const instructions: IrInstruction[] = [];
    const labelMap = new Map<string, number>();
    const pending: PendingPatch[] = [];

    for (const line of program.lines) {
      if (line.label) {
        const key = line.label.name.toLowerCase();
        if (labelMap.has(key)) {
          diagnostics.reportError(
            diagnosticCode.duplicateLabel,
            `Duplicate label '${line.label.name}'.`,
            line.label.span
          );
        } else {
          labelMap.set(key, instructions.length);
        }
      }

      for (const statement of line.statements) {
        emitStatement(statement, instructions, pending);
      }
    }

    for (const patch of pending) {
      const target = labelMap.get(patch.labelName.toLowerCase());
      if (target === undefined) {
        diagnostics.report(
          diagnosticCode.unknownLabel,
          "warning",
          `Unknown label '${patch.labelName}'.`,
          patch.span
        );
        continue;
      }

      instructions[patch.instructionIndex] = patch.isCall
        ? { kind: "CallInstruction", targetIndex: target, span: patch.span }
        : { kind: "JumpInstruction", targetIndex: target, span: patch.span };
    }

    return {
      instructions,
      labelMap,
      diagnostics: diagnostics.items
    };
  }
}

function emitStatement(
  statement: StatementSyntax,
  instructions: IrInstruction[],
  pending: PendingPatch[]
): void {
  switch (statement.kind) {
    case "AssignmentStatement":
      instructions.push({
        kind: "AssignInstruction",
        name: statement.name,
        expression: statement.expression,
        forceString: false,
        span: statement.span
      });
      return;
    case "InstrStatement":
      instructions.push({
        kind: "AssignInstruction",
        name: statement.name,
        expression: statement.expression,
        forceString: true,
        span: statement.span
      });
      return;
    case "PrintStatement":
      instructions.push({
        kind: "PrintInstruction",
        textExpression: statement.textExpression,
        appendNewLine: statement.appendNewLine,
        span: statement.span
      });
      return;
    case "BtnStatement":
      instructions.push({
        kind: "AddButtonInstruction",
        targetExpression: statement.targetExpression,
        captionExpression: statement.captionExpression,
        span: statement.span
      });
      return;
    case "PerkillStatement":
      instructions.push({ kind: "PerkillInstruction", span: statement.span });
      return;
    case "InvkillStatement":
      instructions.push({
        kind: "InvkillInstruction",
        itemExpression: statement.itemExpression,
        span: statement.span
      });
      return;
    case "InvAddStatement":
      instructions.push({
        kind: "InvAddInstruction",
        countExpression: statement.countExpression,
        itemExpression: statement.itemExpression,
        span: statement.span
      });
      return;
    case "InvRemoveStatement":
      instructions.push({
        kind: "InvRemoveInstruction",
        countExpression: statement.countExpression,
        itemExpression: statement.itemExpression,
        span: statement.span
      });
      return;
    case "EndStatement":
      instructions.push({ kind: "ReturnOrHaltInstruction", span: statement.span });
      return;
    case "GotoStatement": {
      const label = tryStaticLabel(statement.target);
      if (label) {
        const instructionIndex = instructions.length;
        instructions.push({ kind: "JumpInstruction", targetIndex: -1, span: statement.span });
        pending.push({ instructionIndex, labelName: label, isCall: false, span: statement.span });
      } else {
        instructions.push({
          kind: "JumpDynamicInstruction",
          targetExpression: statement.target,
          span: statement.span
        });
      }
      return;
    }
    case "ProcStatement": {
      const label = tryStaticLabel(statement.target);
      if (label) {
        const instructionIndex = instructions.length;
        instructions.push({ kind: "CallInstruction", targetIndex: -1, span: statement.span });
        pending.push({ instructionIndex, labelName: label, isCall: true, span: statement.span });
      } else {
        instructions.push({
          kind: "CallDynamicInstruction",
          targetExpression: statement.target,
          span: statement.span
        });
      }
      return;
    }
    case "IfStatement":
      emitIf(statement, instructions, pending);
      return;
    default:
      instructions.push({ kind: "NoOpInstruction", span: statement.span });
  }
}

function emitIf(
  statement: Extract<StatementSyntax, { kind: "IfStatement" }>,
  instructions: IrInstruction[],
  pending: PendingPatch[]
): void {
  const jumpFalseIndex = instructions.length;
  instructions.push({
    kind: "JumpIfFalseInstruction",
    conditionExpression: statement.condition,
    targetIndex: -1,
    span: statement.span
  });

  for (const item of statement.thenStatements) {
    emitStatement(item, instructions, pending);
  }

  if (statement.elseStatements && statement.elseStatements.length > 0) {
    const jumpEndIndex = instructions.length;
    instructions.push({ kind: "JumpInstruction", targetIndex: -1, span: statement.span });

    const elseStart = instructions.length;
    instructions[jumpFalseIndex] = {
      kind: "JumpIfFalseInstruction",
      conditionExpression: statement.condition,
      targetIndex: elseStart,
      span: statement.span
    };

    for (const item of statement.elseStatements) {
      emitStatement(item, instructions, pending);
    }

    instructions[jumpEndIndex] = {
      kind: "JumpInstruction",
      targetIndex: instructions.length,
      span: statement.span
    };
  } else {
    instructions[jumpFalseIndex] = {
      kind: "JumpIfFalseInstruction",
      conditionExpression: statement.condition,
      targetIndex: instructions.length,
      span: statement.span
    };
  }
}

function tryStaticLabel(expression: ExpressionSyntax): string | null {
  switch (expression.kind) {
    case "IdentifierExpression":
      return expression.name.trim() || null;
    case "StringLiteralExpression":
      return expression.value.trim() || null;
    default:
      return null;
  }
}
